File: packages/toolcraft-design-rust/src/acp-events.js
import { createRequire } from "node:module";
import { color } from "./color.js";
import { resolveOutputFormat } from "./logging.js";
import { getAcpWriter } from "./acp-writer.js";
const native = createRequire(import.meta.url)("./toolcraft-design-rust.node");
function writeEvent(event, format, first, second = "", cached = "", cost = "") {
  const [style, text] = native.designAcpEvent(
    event,
    format === "markdown",
    first,
    second,
    cached,
    cost
  );
  getAcpWriter()(style ? color[style](text) : text);
}
export function renderToolStart(kind, title) {
  const format = resolveOutputFormat();
  if (format === "json") getAcpWriter()(JSON.stringify({ event: "tool_start", kind, title }));
  else writeEvent("tool_start", format, kind, title);
}
export function renderToolComplete(kind) {
  const format = resolveOutputFormat();
  if (format === "json") getAcpWriter()(JSON.stringify({ event: "tool_complete", kind }));
  else writeEvent("tool_complete", format, kind);
}
export function renderReasoning(text) {
  const format = resolveOutputFormat();
  if (format === "json") getAcpWriter()(JSON.stringify({ event: "reasoning", text }));
  else writeEvent("reasoning", format, text);
}
export function renderError(message) {
  const format = resolveOutputFormat();
  if (format === "json") getAcpWriter()(JSON.stringify({ event: "error", message }));
  else writeEvent("error", format, message);
}
export function renderPermissionRejected(title) {
  const format = resolveOutputFormat();
  if (format === "json") getAcpWriter()(JSON.stringify({ event: "permission_rejected", title }));
  else writeEvent("permission_rejected", format, title);
}
export function renderUsage(tokens) {
  const format = resolveOutputFormat();
  const cached =
    typeof tokens.cached === "number" && tokens.cached > 0 ? ` (${tokens.cached} cached)` : "";
  let cost = "";
  if (typeof tokens.costUsd === "number")
    cost = ` (${new Intl.NumberFormat("en-US", { style: "currency", currency: "USD", minimumFractionDigits: 2, maximumFractionDigits: 6 }).format(tokens.costUsd)})`;
  if (format === "json") {
    getAcpWriter()(
      JSON.stringify({
        event: "usage",
        inputTokens: tokens.input,
        outputTokens: tokens.output,
        cachedTokens: tokens.cached ?? 0,
        costUsd: tokens.costUsd ?? 0
      })
    );
    return;
  }
  if (format !== "markdown") getAcpWriter()("");
  writeEvent("usage", format, `${tokens.input}`, `${tokens.output}`, cached, cost);
}

File: packages/toolcraft-design-rust/src/acp_events.rs
//! ACP event text policies; host retains writer and structured-object observations.
pub struct Event {
    pub style: &'static str,
    pub text: Vec<u16>,
}
fn u(text: &str) -> Vec<u16> {
    text.encode_utf16().collect()
}
fn join(parts: &[&[u16]]) -> Vec<u16> {
    parts.concat()
}
fn kind_color(kind: &[u16]) -> &'static str {
    if kind.iter().copied().eq("exec".encode_utf16()) {
        "yellow"
    } else if kind.iter().copied().eq("edit".encode_utf16()) {
        "magenta"
    } else if kind.iter().copied().eq("read".encode_utf16()) {
        "cyan"
    } else if kind.iter().copied().eq("search".encode_utf16()) {
        "blue"
    } else {
        "dim"
    }
}
pub fn render(
    event: &str,
    markdown: bool,
    first: &[u16],
    second: &[u16],
    cached: &[u16],
    cost: &[u16],
) -> Event {
    let (style, text) = match event {
        "tool_start" => {
            if markdown {
                ("", join(&[&u("- *→ "), first, &u(": "), second, &u("*")]))
            } else {
                (
                    kind_color(first),
                    join(&[&u("  → "), first, &u(": "), second]),
                )
            }
        }
        "tool_complete" => {
            if markdown {
                ("", join(&[&u("- *✓ "), first, &u("*")]))
            } else {
                (kind_color(first), join(&[&u("  ✓ "), first]))
            }
        }
        "reasoning" => {
            let short = if first.len() <= 80 {
                first.to_vec()
            } else {
                join(&[&first[..77], &u("...")])
            };
            if markdown {
                ("", join(&[&u("- *thinking:* "), &short]))
            } else {
                ("dim", join(&[&u("  · "), &short]))
            }
        }
        "usage" => {
            if markdown {
                (
                    "",
                    join(&[
                        &u("- **tokens:** "),
                        first,
                        &u(" in → "),
                        second,
                        &u(" out"),
                        cost,
                    ]),
                )
            } else {
                (
                    "dim",
                    join(&[
                        &u("· tokens: "),
                        first,
                        &u(" in"),
                        cached,
                        &u(" → "),
                        second,
                        &u(" out"),
                        cost,
                    ]),
                )
            }
        }
        "error" => {
            if markdown {
                ("", join(&[&u("- **error:** "), first]))
            } else {
                ("red", join(&[&u("✗ "), first]))
            }
        }
        "permission_rejected" => {
            if markdown {
                ("", join(&[&u("- **permission rejected:** "), first]))
            } else {
                ("yellow", join(&[&u("  ✗ permission rejected: "), first]))
            }
        }
        _ => ("", Vec::new()),
    };
    Event { style, text }
}
